providers/adfs.go

```diff
@@ -0,0 +1,191 @@
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		params.Add("resource", p.ProtectedResource.String())
+	}
+	params.Add("grant_type", "authorization_code")
+	params.Add("code", code)
+	params.Add("redirect_uri", redirectURL)
+
+	err = requests.New(p.RedeemURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		WithBody(bytes.NewBufferString(params.Encode())).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Do().
+		UnmarshalInto(&jsonResponse)
+	if err != nil {
+		return
+	}
+
+	created := time.Now()
+	expires := time.Unix(jsonResponse.ExpiresIn, 0)
+
+	return &sessions.SessionState{
+		AccessToken:  jsonResponse.AccessToken,
+		IDToken:      jsonResponse.IDToken,
+		CreatedAt:    &created,
+		ExpiresOn:    &expires,
+		RefreshToken: jsonResponse.RefreshToken,
+	}, nil
+}
+
+// EnrichSession to add email
+func (p *AdfsProvider) EnrichSession(_ context.Context, s *sessions.SessionState) error {
+	c, err := adfsClaimsFromToken(s.IDToken)
+	if err != nil {
+		return err
+	}
+
+	s.Email = c.Email
+	s.User = c.Upn
+	return nil
+}
+
+// RefreshSessionIfNeeded checks if the session has expired and uses the
+// RefreshToken to fetch a new ID token if required
+func (p *AdfsProvider) RefreshSessionIfNeeded(ctx context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil || s.ExpiresOn.After(time.Now()) || s.RefreshToken == "" {
+		return false, nil
+	}
+
+	origExpiration := s.ExpiresOn
+
+	err := p.redeemRefreshToken(ctx, s)
+	if err != nil {
+		return false, fmt.Errorf("unable to redeem refresh token: %v", err)
+	}
+
+	logger.Printf("refreshed id token %s (expired on %s)\n", s, origExpiration)
+	return true, nil
+}
+
+func (p *AdfsProvider) redeemRefreshToken(ctx context.Context, s *sessions.SessionState) (err error) {
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", p.ClientSecret)
+	params.Add("refresh_token", s.RefreshToken)
+	params.Add("grant_type", "refresh_token")
+
+	var jsonResponse struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in,string"`
+		IDToken      string `json:"id_token"`
+	}
+
+	err = requests.New(p.RedeemURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		WithBody(bytes.NewBufferString(params.Encode())).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Do().
+		UnmarshalInto(&jsonResponse)
+
+	if err != nil {
+		return
+	}
+
+	now := time.Now()
+	expires := time.Unix(jsonResponse.ExpiresIn, 0)
+	s.AccessToken = jsonResponse.AccessToken
+	s.IDToken = jsonResponse.IDToken
+	s.RefreshToken = jsonResponse.RefreshToken
+	s.CreatedAt = &now
+	s.ExpiresOn = &expires
+	return
+}
+
+func adfsClaimsFromToken(token string) (*adfsClaims, error) {
+	jwt := strings.Split(token, ".")
+	if len(jwt) != 3 {
+		return nil, fmt.Errorf("failed to parse token, wrong format")
+	}
+	jwtData := strings.TrimSuffix(jwt[1], "=")
```

Comment:
If you end up keeping this method, we'll need to refactor this: we validate all token signatures & don't trust any payload claims unless we successfully validate the token.

You would need to do something like this:
https://github.com/oauth2-proxy/oauth2-proxy/blob/d2ffef2c7e0a622295c42cb2fdd70ef14593d03c/providers/oidc.go#L182

With an `oidc.IDTokenVerify` that gets initialized from your OIDC Provider's discovery document (`issuer_url + /.well-known/openid-configuration`). The OIDCProvider does this when it is initialized.